File: reference/spring-boot-java-gradle-hexagonal/src/main/java/com/example/demo/application/port/inbound/ManageWidgets.java
package com.example.demo.application.port.inbound;

import com.example.demo.domain.Widget;
import java.util.List;

/**
 * What the application can be asked to do, stated without reference to who asks.
 *
 * <p>{@code inbound}/{@code outbound} rather than hexagonal's usual {@code in}/{@code out}: the
 * Kotlin backend emits the same structure, {@code in} is a Kotlin hard keyword, and the only way
 * to spell it as a package segment there is backticks — which ktlint rejects. Java would have
 * taken {@code in} happily; the choice is made once, for both languages, rather than letting two
 * trees for the same architecture diverge on a detail nobody would remember.
 *
 * <p>A driving port. The web adapter depends on this interface and not on the class that
 * implements it, which is what makes "the HTTP layer is one of several possible front doors" a
 * structural fact rather than an aspiration — a CLI or a message consumer is a second adapter
 * against this same interface and no change to anything behind it.
 */
public interface ManageWidgets {

    List<Widget> findAll();

    Widget findById(long id);

    Widget create(String name, int quantity);

    Widget restock(long id, int amount);

    void delete(long id);
}

File: reference/spring-boot-kotlin-gradle-hexagonal/src/main/kotlin/com/example/demo/application/port/inbound/ManageWidgets.kt
package com.example.demo.application.port.inbound

import com.example.demo.domain.Widget

/**
 * What the application can be asked to do, stated without reference to who asks.
 *
 * A driving port. The web adapter depends on this interface and not on the class that implements
 * it, which is what makes "the HTTP layer is one of several possible front doors" a structural
 * fact rather than an aspiration.
 *
 * `inbound`/`outbound` rather than hexagonal's usual `in`/`out`, and the reason is this file's
 * language: `in` is a Kotlin hard keyword, and the only way to spell it as a package segment is
 * backticks, which ktlint rejects outright. Java would have taken `in` happily — which is exactly
 * why the choice is made once, for both languages, instead of letting the two trees diverge on a
 * detail nobody would remember.
 */
interface ManageWidgets {
    fun findAll(): List<Widget>

    fun findById(id: Long): Widget

    fun create(
        name: String,
        quantity: Int,
    ): Widget

    fun restock(
        id: Long,
        amount: Int,
    ): Widget

    fun delete(id: Long)
}
